### DISCUSSION TITLE
How to make a second request to the provider's API to fetch an image

### DISCUSSION THREAD

**Discussion by italodeandra** [nextauthjs/next-auth#853] [Help]
**Your question**
I would like to know if its possible to request the image or if we can get the token that NextAuth that used to fetch the user info.

**What are you trying to do**
I'm currently implementing a custom provider MSAL. It is working perfectly, but the image from Microsoft accounts are a separate request to their Graph API `https://graph.microsoft.com/v1.0/me/photo/$value`. 

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [x] Found the documentation helpful
* [x] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [x] Found the example project helpful
* [ ] Did not find the example project helpful

**Comment by jachinte** [comment]
@italodeandra Since the `session` callback returns a promise, I think you could run your extra request there.

**Comment by italodeandra** [comment]
@jachinte That's nice to know but still doesn't fix my problem. Because the Graph API is only accessible with the token. The way I foresee this happening is getting access to NextAuth API inside the `session` callback, so I can call another endpoint. But I don't think this is possible now. Do you have any suggestion?

**Comment by jachinte** [comment]
@italodeandra You can get the token in the `jwt` callback. It's included in the `account` argument once the user logs in.

**Comment by robertjdominguez** [comment]
@italodeandra -- would you be willing to share your MSAL provider config? Starting a project that requires authentication via this route and it would be helpful to have a working example to reference.

**Comment by balazsorban44** [comment]
so we recently added `tokens` as the second parameter to the provider profile callback. you can use the access token (or ID token I guess) inside there to fetch more information.

https://next-auth.js.org/configuration/providers#using-a-custom-provider

I wouldn't recommend using the `session` callback, as it would make too much unnecessary requests. You can expand the session the first time the jwt callback is called. https://next-auth.js.org/configuration/callbacks#jwt-callback
